### PR TITLE
[WIP] Implement Countable where possible

### PR DESCRIPTION
**Pull request by tgalopin**
As proposed in #7 and following the same pattern as #9, here is my PR for countable stores.

Agenda:

- Create a `CountableStore` interface for countable stores ✔
- Create a `CountableDecorator` for counting others ✔
- Implement `count()` for native countable stores ✔
- Create tests for all of that ✔
- Squash the commits once finished **WIP**

I also created an `AbstractDelegatedDecorator` and a `AbstractDelegatingDecoratorTest` to avoid code duplication.

### CODE REVIEW

**Comment by tgalopin** [comment]
Could you relaunch the Scrunitizer analysis? It seems to have failed without valid reason.

File: src/AbstractDelegatingDecorator.php [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+ *
+ * It is used by decorators that need to override only a few specific
+ * methods (such as SortableDecorator or CountableDecorator).
+ *
+ * @since  1.0
+ *
+ * @author Bernhard Schussek <bschussek@gmail.com>
+ * @author Titouan Galopin <galopintitouan@gmail.com>
+ */
+class AbstractDelegatingDecorator implements KeyValueStore
```
Comment (webmozart) [inline src/AbstractDelegatingDecorator.php]: A decorator delegates to the decorated class by definition, so I think you can simplify the name to `AbstractDecorator`.

File: src/CountableDecorator.php [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+     * @var int
+     */
+    private $cache;
+
+    /**
+     * Is the cache fresh enough to be served?
+     *
+     * @var bool
+     */
+    private $cachIsFresh = false;
```
Comment (webmozart) [inline src/CountableDecorator.php]: `$cacheIsFresh`

File: src/AbstractDelegatingDecorator.php (hunk 1 above)
Comment (tgalopin) [inline src/AbstractDelegatingDecorator.php]: Right :) .

File: src/CountableDecorator.php (hunk 2 above)
Comment (tgalopin) [inline src/CountableDecorator.php]: Oups, thanks!

**Comment by webmozart** [comment]
Looks good! Could you fix the remaining issues and squash your commits please? Thanks @tgalopin :)

**Comment by tgalopin** [comment]
I think it's ready now :) .

**Comment by webmozart** [comment]
Cool, thanks :)

**Pull request merged**